File: oldOnes/python/Old/mergeSort.py
def merge_sort(arr):
    def sort(left, right):
        if right - left < 2:
            return
        mid = (left + right) // 2
        sort(left, mid)
        sort(mid, right)
        merge(left, mid, right)

    def merge(left, mid, right):
        temp = []
        l, h = left, mid

        while l < mid and h < right:
            if arr[l] < arr[h]:
                temp.append(arr[l])
                l += 1
            else:
                temp.append(arr[h])
                h += 1

        while l < mid:
            temp.append(arr[l])
            l += 1
        while h < right:
            temp.append(arr[h])
            h += 1
        
        for i in range(left, right):
            arr[i] = temp[i - left]
            
    return sort(0, len(arr))


        

File: oldOnes/java/src/MergeSort.java
public class MergeSort {

    int[] mergeSort(int[] arr){
        int[] tmp = new int[arr.length];
        mergeSort(arr, tmp, 0, arr.length - 1);
        return arr;
    }

    void mergeSort(int[] arr, int[] tmp, int left, int right){
        if(left < right){
            int mid = (left + right) / 2;
            mergeSort(arr, tmp, left, mid);
            mergeSort(arr, tmp, mid + 1, right);
            merge(arr, tmp, left, mid, right);
        }
    }
    void merge(int[] arr, int[] tmp, int left, int mid, int right){
        for(int i = left; i <= right; i++){
            tmp[i] = arr[i];
        }
        int pi1 = left;
        int pi2 = mid + 1;
        int index = left;

        while (pi1 <= mid && pi2 <= right){
            if (tmp[pi1] <= tmp[pi2]){
                arr[index] = tmp[pi1];
                pi1++;
            }else{
                arr[index] = tmp[pi2];
                pi2++;
            }
            index++;
        }
        for(int i = 0; i <= mid - pi1; i++){
            arr[index + 1] = tmp[pi1 + 1];
        }
    }

    public static void main(String[] args) {
        int[] arr = {4, 3, 2, 1};
        MergeSort m = new MergeSort();
        int[] sorted_arr = m.mergeSort(arr);
        for (int e: sorted_arr) {
            System.out.println(e);
        }
    }
}
